src/miral/input_device_config.cpp

```diff
@@ -62,56 +78,187 @@ char const* const touchpad_click_mode_opt= "touchpad-click-mode";
 
 char const* const touchpad_click_mode_none = "none";
 char const* const touchpad_click_mode_area = "area";
-char const* const touchpad_click_mode_finger_count = "finger-count";
+char const* const touchpad_click_mode_clickfinger = "clickfinger";
+char const* const touchpad_middle_mouse_button_emulation_opt= "touchpad-middle-mouse-button-emulation";
 
 class InputDeviceConfig : public mi::InputDeviceObserver
 {
 public:
-    InputDeviceConfig(std::optional<bool> disable_while_typing,
-                      std::optional<bool> tap_to_click,
-                      std::optional<MirPointerAcceleration> mouse_cursor_acceleration,
-                      std::optional<double> mouse_cursor_acceleration_bias,
-                      std::optional<double> mouse_scroll_speed_scale,
-                      std::optional<MirPointerAcceleration> touchpad_cursor_acceleration,
-                      std::optional<double> touchpad_cursor_acceleration_bias,
-                      std::optional<double> touchpad_scroll_speed_scale,
-                      std::optional<MirTouchpadClickMode> click_mode,
-                      std::optional<MirTouchpadScrollMode> scroll_mode);
+    explicit InputDeviceConfig(std::shared_ptr<mir::options::Option> const& options);
     void device_added(std::shared_ptr<mi::Device> const& device) override;
     void device_changed(std::shared_ptr<mi::Device> const&) override {}
     void device_removed(std::shared_ptr<mi::Device> const&) override {}
     void changes_complete() override {}
 private:
     std::optional<bool> const disable_while_typing;
+    std::optional<bool> const disable_with_external_mouse;
+    std::optional<MirPointerHandedness> const mouse_handedness;
     std::optional<MirPointerAcceleration> const mouse_cursor_acceleration;
     std::optional<double> const mouse_cursor_acceleration_bias;
-    std::optional<double> const mouse_scroll_speed_scale;
+    std::optional<double> const mouse_vscroll_speed;
+    std::optional<double> const mouse_hscroll_speed;
     std::optional<MirPointerAcceleration> const touchpad_cursor_acceleration;
     std::optional<double> const touchpad_cursor_acceleration_bias;
-    std::optional<double> const touchpad_scroll_speed_scale;
+    std::optional<double> const touchpad_vscroll_speed;
+    std::optional<double> const touchpad_hscroll_speed;
     std::optional<MirTouchpadClickMode> const click_mode;
     std::optional<MirTouchpadScrollMode> const scroll_mode;
     std::optional<bool> const tap_to_click;
+    std::optional<bool> const middle_mouse_button_emulation;
 };
 
+auto clamp_to_range(std::optional<double> opt_val)-> std::optional<double>
```

Comment:
Preexisting, but while we're here... maybe this should mention what the range is? Or accept the range?

Probably `clamp_to_range(std::optional<double> opt_val, double min, double max)` would be clearest?